appveyor.yml

```diff
@@ -6,16 +6,21 @@ image: Visual Studio 2017
 
 clone_folder: c:\gopath\src\github.com\urfave\cli
 
+cache:
+  - node_modules
+
 environment:
   GOPATH: C:\gopath
   GOVERSION: 1.11.x
+  GO111MODULE: on
+  GOPROXY: https://proxy.golang.org
 
 install:
   - set PATH=%GOPATH%\bin;C:\go\bin;%PATH%
   - go version
   - go env
-  - go get github.com/urfave/gfmrun/...
-  - go get -v -t ./...
+  - go get github.com/urfave/gfmrun/cmd/gfmrun
+  - go mod vendor
```

Comment:
❓ _(blocking question)_ Not sure I understand this line of diff?

Reply:
The older lines work with Go Modules disabled. With Go Modules enabled, the tool needs to locate a package with a tag and complete path. Hence the change.

`go mod vendor` is the new command to download dependencies.

Reply:
I'm reading this https://github.com/golang/go/wiki/Modules and it looks like it says that `go mod vendor` is only needed if you're using vendoring?

So I think I may be missing something, because it looks like `go get` is what you use for downloading dependencies. If you even need to run `go get` that is, since that page above says

> Note there was no go get required in the example above

Here https://github.com/golang/go/wiki/Modules#daily-workflow

Reply:
This is different from the daily workflow where you already have a vendor directory and you might want to download a single new dependency or update some of the existing dependencies. But in the CI environment when we clone the repo, there is no vendor folder. Hence `go mod vendor` is required to initialize the vendor folder and get all the dependencies in place.

Reply:
Also as seen here:

> Standard commands like go build or go test will automatically add new dependencies as needed to satisfy imports (updating go.mod and downloading the new dependencies)

I don't want to risk updating any dependency. I just want to download what has been mentioned in go.mod and go.sum and proceed with testing

Reply:
> But in the CI environment when we clone the repo, there is no vendor folder. Hence go mod vendor is required to initialize the vendor folder and get all the dependencies in place.

I'm still not following? There shouldn't be any vendor folder involved here? The go modules system completely replaces the usage of the vendor folder.

> I don't want to risk updating any dependency. I just want to download what has been mentioned in go.mod and go.sum and proceed with testing

The thing I've done here (in python and javascript) is write a small test that asserts that the dependency file hasn't changed. That's fairly standard practice, I think.

---

Also, as a meta comment, I'm not sure why there's a diff on this file at all inside of this V2 PR. I think the changes on this specific file would be a bit easier to discuss inside their own PR?

Reply:
These changes are already a part of master when we added go modules support. This is just coming into the v2 branch now.

Reply:
Check .travis.yml The appveyor does not include GO module support on Master

Reply:
> The appveyor does not include GO module support on Master

This makes sense 👍 but I would still say my ⭐⭐⭐ _(must change)_ feedback here is that I need the diff for the V2 PR to shrink, which means moving changes like this one into their own PR. I'm going to do that myself now.

Reply:
=> https://github.com/urfave/cli/pull/898

Reply:
`go mod vendor` only populates the `vendor/` directory on the local file system. It's not a requirement to use modules, and it's actually ignored unless explicitly opted into for each `go` command.

From [the official documentation](https://github.com/golang/go/wiki/Modules#how-do-i-use-vendoring-with-modules-is-vendoring-going-away):
> By default, go commands like `go build` ignore the vendor directory when in module mode.

There's some details in those docs on why vendoring can be valuable, but I don't think it's providing any benefit in this case. Based on the conversation above, what we're probably looking for here is running `go test` (or other commands) with the `-mod=readonly` flag.

From `go help modules`:

> If invoked with -mod=readonly, the go command is disallowed from the implicit
automatic updating of go.mod described above. Instead, it fails when any changes
to go.mod are needed. This setting is most useful to check that go.mod does
not need updates, such as in a continuous integration and testing system.
The "go get" command remains permitted to update go.mod even with -mod=readonly,
and the "go mod" commands do not take the -mod flag (or any other build flags).

To guarantee that it's used consistently, the `GOFLAGS` env var can be set in the `travis.yml`:
```yaml
environment:
  # ...
  GO111MODULE: on
  GOFLAGS: -mod=vendor
```

Reply:
Big 👍  on `readonly`, I'll stand up a PR for that change in a bit.

Reply:
=> https://github.com/urfave/cli/pull/903

Reply:
Setting `GOFLAGS=-mod=readonly` is causing builds to fail on Travis. However, with regard to above discussion, I have removed the `go mod vendor` and have instead added `go mod tidy`

`go mod tidy` prunes any no-longer-needed dependencies (which are added when we try to install gfmrun and goimports for go vet) from go.mod and add any dependencies needed for other combinations of OS, architecture, and build tags as described [here](https://github.com/golang/go/wiki/Modules#daily-workflow)

CI builds are working now...

Reply:
What really happens here is that when we have `GOFLAGS=-mod=readonly` it prevents the `go.mod` file from being modified. This, however, causes a problem when `gfmrun` tries to execute the examples from `README.md`. It tries to download the `urfave/cli` dependency and modify the blank mod file created with the example with the dependency. Setting mod to read-only causes this to fail.

The workaround here is to add a flag to the `gfmrun` tool that would allow us to specify dependencies for running the examples or ignore the mod readonly for the while.